File: medium/javascript/0649-dota2-senate.js
/**
 * @param {string} senate
 * @return {string}
 */
var predictPartyVictory = function (senate) {
  const queue = [];

  for (let i = 0; i < senate.length; i++) {
    queue.push(senate[i]);
  }

  const direQueue = [];
  const radiantQueue = [];

  while (true) {
    if (direQueue.length >= queue.length) {
      return "Dire";
    } else if (radiantQueue.length >= queue.length) {
      return "Radiant";
    }
    const current = queue.shift();

    if (current === "R") {
      if (direQueue.shift() === undefined) {
        radiantQueue.push("R");
        queue.push("R");
      }
    } else if (current === "D") {
      if (radiantQueue.shift() === undefined) {
        direQueue.push("D");
        queue.push("D");
      }
    }
  }
};

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

/**
 * @param {string} senate
 * @return {string}
 */
var predictPartyVictory = function (senate) {
  const queue = [];

  for (let i = 0; i < senate.length; i++) {
    queue.push(senate[i]);
  }

  let direQueued = 0;
  let radiantQueued = 0;

  while (true) {
    if (direQueued >= queue.length) {
      return "Dire";
    } else if (radiantQueued >= queue.length) {
      return "Radiant";
    }
    const current = queue.shift();

    if (current === "R") {
      if (direQueued === 0) {
        radiantQueued++;
        queue.push("R");
      } else {
        direQueued--;
      }
    } else if (current === "D") {
      if (radiantQueued === 0) {
        direQueued++;
        queue.push("D");
      } else {
        radiantQueued--;
      }
    }
  }
};

File: medium/go/0649-dota2-senate.go
package main

func predictPartyVictory(senate string) string {
	queue := []rune{}

	for _, v := range senate {
		queue = append(queue, v)
	}

	direQueued := 0
	radiantQueued := 0

	for {
		if direQueued >= len(queue) {
			return "Dire"
		} else if radiantQueued >= len(queue) {
			return "Radiant"
		}

		current := queue[0]
		queue = queue[1:]

		if current == 'R' {
			if direQueued == 0 {
				radiantQueued++
				queue = append(queue, 'R')
			} else {
				direQueued--
			}
		} else if current == 'D' {
			if radiantQueued == 0 {
				direQueued++
				queue = append(queue, 'D')
			} else {
				radiantQueued--
			}
		}
	}
}